ray/adaptdl_ray/tune/adaptdl_trainable.py

```diff
@@ -0,0 +1,137 @@
+                       new=P.setup_process_group), \
+                 patch(target='ray.tune.integration.torch.wrap_function',
+                       new=P.wrap_function_patched):
+                _TorchTrainable.setup(self, config)
+
+        # Override the default resources and use custom PG factory
+        @classmethod
+        def default_resource_request(cls, config: Dict) -> Resources:
+            return None
+
+        def get_sched_hints(self):
+            return ray.get(self.workers[0].get_sched_hints.remote())
+
+        def save_all_states(self, trial_state):
+            return ray.get(self.workers[0].save_all_states.remote(trial_state))
+
+        @classmethod
+        def default_process_group_parameters(self) -> Dict:
+            return dict(timeout=timedelta(timeout_s), backend=backend)
+
+    AdaptDLTrainable._function = func
+    AdaptDLTrainable._num_workers = num_workers
+    # Set number of GPUs if we're using them, this is later used when spawning
+    # the trial actors
+    if config.default_device() == "GPU":
+        AdaptDLTrainable._num_gpus_per_worker = 1
+    else:
+        AdaptDLTrainable._num_gpus_per_worker = 0
+
+    # Trainables are named after number of replicas they spawn. This is
+    # essential to associate the right Trainable with the right Trial and PG.
+    AdaptDLTrainable.__name__ = AdaptDLTrainable.__name__.split("_")[0] + \
+        f"_{num_workers}" + f"_{group}"
+    register_trainable(AdaptDLTrainable.__name__, AdaptDLTrainable)
+    return AdaptDLTrainable
+
+
+# To support unit tests and integration tests
+def _train_simple(config: Dict, checkpoint_dir: Optional[str] = None):
+    import torch
+    import torch.nn as nn
+    import torch.optim as optim
+    import adaptdl.torch as adl
+    from ray import tune
+
+    class MyDataset:
+        def __init__(self, xs, ys):
+            self.xs = xs
+            self.ys = ys
+
+        def __getitem__(self, i):
+            return self.xs[i], self.ys[i]
+
+        def __len__(self):
+            return len(self.xs)
+
+    # N is batch size; D_in is input dimension;
+    # H is hidden dimension; D_out is output dimension.
+    N, D_in, H, D_out = 64, 5, 5, 5
+    dataset = MyDataset(torch.randn(N, D_in), torch.randn(N, D_out))
+
+    H = config.get("H", 16)
+    N = config.get("N", 16)
+
+    # Create random Tensors to hold inputs and outputs
+    dataloader = adl.AdaptiveDataLoader(dataset, batch_size=N)
+    dataloader.autoscale_batch_size(4096, local_bsz_bounds=(16, 1024))
+
+    loss_fn = nn.MSELoss()
+
+    # Use the nn package to define our model and loss function.
+    model = torch.nn.Sequential(
+        torch.nn.Linear(D_in, H),
+        torch.nn.ReLU(),
+        torch.nn.Linear(H, D_out),
+    )
+    optimizer = optim.SGD(model.parameters(), lr=0.1)
+
+    model = adl.AdaptiveDataParallel(model, optimizer)
+
+    loss = torch.Tensor([0.0])
+    for epoch in adl.remaining_epochs_until(config.get("epochs", 10)):
```

Comment:
With relatively expensive tests like this, we should probably have a separate job for testing the /ray/adaptdl_ray directory so it can run in parallel with the k8s and client tests. Those take enough time already.

Reply:
This isn't a blocking comment, btw